### PR TITLE
Robust parameter generation & URL date selector for Sites

### PR DESCRIPTION
**Pull request by tleraj**
The purpose of this PR is to close
 - [x] https://github.com/aqualinkorg/aqualink-app/issues/637
 - [x] https://github.com/aqualinkorg/aqualink-app/issues/640

TODOs:
 - [x] Fix API tests
 - [x] Enhance `ParseDatePipe`

### CODE REVIEW

File: packages/api/src/utils/time-series.utils.ts [hunk 1]
```diff
@@ -59,13 +59,19 @@ export const groupByMetricAndSource = <T extends TimeSeriesGroupable>(
...
-  metrics: Metric[],
-  hourly: boolean,
   siteId: number,
+  metrics: Metric[],
+  start?: Date,
+  end?: Date,
+  hourly?: boolean,
   surveyPointId?: number,
 ): Promise<TimeSeriesData[]> => {
+  const endDate = end && !isNaN(end.getTime()) ? end : new Date();
```
Comment (ericboucher) [inline packages/api/src/utils/time-series.utils.ts:69]: add a comment

Comment (tleraj) [reply packages/api/src/utils/time-series.utils.ts]: Added this logic into a util function in https://github.com/aqualinkorg/aqualink-app/pull/642/commits/81af6a46eea5e4c021d9438435d39a925486f3a0 :wink:

File: packages/website/src/helpers/siteUtils.ts [hunk 2]
```diff
@@ -71,3 +74,26 @@ export const setSiteNameFromList = ({
...
+
+      return `${acc}${specialChar}${key}=${String(value)}`;
+    },
+    ""
+  );
+
+  return `time-series/sites/${siteId}${
+    pointId ? `/site-survey-points/${pointId}` : ""
+  }${stringifiedParams}`;
+};
```
Comment (ericboucher) [inline packages/website/src/helpers/siteUtils.ts:99]: Is there no library that takes a url and an object to return the url with query params? we're not going to do this for every api endpoint...

**Review by ericboucher — CHANGES_REQUESTED** [review]
Looks good overall, just one main comment around query generation

File: packages/website/src/helpers/siteUtils.ts (hunk 2 above)
Comment (ericboucher) [reply packages/website/src/helpers/siteUtils.ts:99]: Eg. https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams

Comment (ericboucher) [reply packages/website/src/helpers/siteUtils.ts:99]: Or other options listed in https://stackoverflow.com/questions/111529/how-to-create-query-parameters-in-javascript

Comment (tleraj) [reply packages/website/src/helpers/siteUtils.ts]: Using `URLSearchParams` as of https://github.com/aqualinkorg/aqualink-app/pull/642/commits/713a4f6edb8a28bae15b631bb2e9697e7e225b6f

File: packages/website/src/helpers/siteUtils.ts [hunk 3]
```diff
@@ -1,8 +1,11 @@
 import { LatLng } from "leaflet";
-import { maxBy, meanBy } from "lodash";
+import { maxBy, meanBy, omitBy, isUndefined, isArray } from "lodash";
 
-import { longDHW } from "../store/Sites/helpers";
 import { Site, UpdateSiteNameFromListArgs } from "../store/Sites/types";
+import type { TimeSeriesDataRequestParams } from "../store/Sites/types";
+
+export const longDHW = (dhw: number | null): string =>
```
[the remark this answers is not in the document]
Comment (tleraj) [reply packages/website/src/helpers/siteUtils.ts:7]: Adding the `constructTimeSeriesDataRequestUrl` util into `packages/website/src/helpers/siteUtils.ts` caused a dependency cycle. That's why I moved this here

File: packages/website/src/helpers/siteUtils.ts [hunk 4]
```diff
@@ -71,3 +74,17 @@ export const setSiteNameFromList = ({
...
 }: UpdateSiteNameFromListArgs): Site[] | undefined =>
   list?.map((item) => (item.id === id && name ? { ...item, name } : item));
+
+export const constructTimeSeriesDataRequestUrl = ({
+  siteId,
+  pointId,
+  ...rest
+}: TimeSeriesDataRequestParams) => {
+  const params = new URLSearchParams({
+    ...omitBy(rest, isUndefined),
```
Comment (ericboucher) [inline packages/website/src/helpers/siteUtils.ts:84]: Let's create a util function like that that we can reuse:

`constructQueryParameters`:

```
const params = new URLSearchParams({
    ...omitBy(rest, isUndefined),
    }).toString();

return params.length ? `?${params}` : "";
```

**Review by ericboucher — APPROVED** [review]

**Pull request merged**